src/content/en/fundamentals/engage-and-retain/push-notifications/handling-messages.markdown

```diff
@@ -0,0 +1,448 @@
+      ]
+    }
+    self.registration.showNotification(data.title, noteOptions);
+  })  
+});
+{% endhighlight %}
+
+## Put actions on the notification {#notification-actions}
+
+We've already seen examples of notifications with actions built into them. Let's 
+look at how they're implemented and how to respond to them.
+
+Recall that `showNotification()` takes an options argument with one or more 
+optional actions.
+
+{% highlight javascript %}
+ServiceWorkerRegistration.showNotification(title, {  
+  body: data.body,  
+  icon: (data.icon ? data.icon : '/images/i_face_black_24dp_2x.png'),  
+  vibrate: [200, 100, 200, 100, 200, 100, 400],  
+  tag: data.tag,  
+  actions: [  
+    {action: 'change', title: 'Ask for reschedule'},  
+    {action: 'confirm', title: 'Confirm'}  
+  ],  
+  data: data  
+})
+{% endhighlight %}
+
+The image below shows what this looks like. The notification says that Stacy has
+confirmed an appointment for 3:00 PM. The recipient can either respond with
+their own confirmation or ask for the appointment to be rescheduled. For the
+former we will send a message directly to the server. For the later we will open
+the application to an appropriate interface.
+
+<img src="images/confirmation.png" width="393" height="200" />
+
+First, let's add a `notificationclick` event handler to the service worker. Also, 
+close the notification.
+
+{% highlight javascript %}
+self.addEventListener('notificationclick', function(event) {  
+  event.notification.close();  
+  // Process the user action.  
+});
+{% endhighlight %}
+
+Next, we need some logic to figure out where the notification was clicked. Did
+the user click Confirm, Reschedule, or neither?
+
+{% highlight javascript %}
+self.addEventListener('notificationclick', function(event) {
+  event.notification.close();
+  if (event.action === 'confirm') {
+    // Send the confirmation to the server.
+  } else if (event.action === 'change') {
+    // Open the application to a place where the user can reschedule.
+  } else {
+    // Just open the app.
+  }
+});
+{% endhighlight %}
+
+If the user clicked confirm, we can send that straight back to the server
+without opening the application (lines 4 through 12).
+
+{% highlight javascript linenos %}
+self.addEventListener('notificationclick', function(event) {
+  event.notification.close();
+  if (event.action === 'confirm')
+    var fetchOptions = {
+      method: 'post',
+      headers: new Headers({
+        'Content-Type': 'application/json'
+      }),
+      body: event.notification.data.confirmation_id
+    };
+    var confirmation = new Request('/back/end/system/confirm');
+    event.waitUntil(fetch(confirmation, fetchOptions));
+  } else if (event.action === 'change') {
+    // Open the application to a place where the user can reschedule.
+  } else {
+    // Just open the app.
+  }
+});
+{% endhighlight %}
+
+For the change and default click options, we want to navigate to different URLs.
```

Comment:
Tweak to "If the user clicks change or the clicks on the notification rather than one of the action buttons, we want to..."